### PR TITLE
feat(contentful): add traced SVGs to Contentful images

### PR DESCRIPTION
**Pull request by axe312ger**
Since Contentful is supported by the SQIP transformer, this now enables traced SVG previews for Contentful Assets :)

### CODE REVIEW

**Comment by axe312ger** [comment]
contentful-traced-svg

**Comment by axe312ger** [comment]
`using-contentful` deployment is failing since it can't find the new fragments. I am pretty sure this is because it is not published yet.

BUT

I am getting the same error on my local installation. The gif above I actually worked because I did not use the fragment, I put the required values there on my own. Any ideas why my Fragments are not found by Gatsby?

```
1:31:33 PM: error GraphQLCompilerContext: Cannot find fragment `GatsbyContentfulSizes_tracedSVG`. Please make sure the fragment exists in `GatsbyContentfulSizes`.
1:31:34 PM: 
1:31:34 PM:   Error: GraphQLCompilerContext: Cannot find fragment `GatsbyContentfulSizes_tra  cedSVG`. Please make sure the fragment exists in `GatsbyContentfulSizes`.
```
https://app.netlify.com/sites/using-contentful/deploys/5b112d828df894600842a8f4

**Comment by m-allanson** [comment]
It's caused by these lines I think:

https://github.com/gatsbyjs/gatsby/blob/31530b855097b43e028a47ecc2292faec821450e/packages/gatsby-source-contentful/src/gatsby-node.js#L214-L217

The fragments are copied from your local Gatsby install's `src` directory into your project's `.cache` directory. When using `gatsby-dev`, it copies only the compiled js files to your local Gatsby install , not the `src` files. 

So the above lines are copying the fragments from the published version of Gatsby instead of your local development version.

**Comment by axe312ger** [comment]
Alright. So I'd say this is ready for review/release? <3

**Comment by KyleAMathews** [comment]
Nice!

**Comment by pieh** [comment]
> The fragments are copied from your local Gatsby install's src directory into your project's .cache directory.

How about copying built fragments.js then? What's the reason for copying it from `src` (which cause problems like this)? (this could also be changed then in `gatsby-transformer-sharp` which copies fragments.js from `src` too)

There is also issue of download images - right now this is duplicated in sqip plugin, and there is PR from @Khaledgarbaya that adds this in yet another place. Problem I have with this sqip and this PR have separate caches and It would be nice to abstract it, so this can be reused in all places I mentioned (not true blocker, but would be nice to solve it)

**Comment by axe312ger** [comment]
> How about copying built fragments.js then? What's the reason for copying it from src (which cause problems like this)? (this could also be changed then in gatsby-transformer-sharp which copies fragments.js from src too)

Makes totally sense. Should be a seperate PR which maybe gets merged before this to fix deployment preview 🤔 

> There is also issue of download images - right now this is duplicated in sqip plugin

I agree, we should have a generic function which all plugins use for this

> Problem I have with this sqip and this PR have separate caches and It would be nice to abstract it, so this can be reused in all places I mentioned (not true blocker, but would be nice to solve it)

Same here, maybe gatsby should define a folder where such files should be cached. Not all downloaded files need to end up in public/static. Maybe `.cache/files/[plugin-name]`. This also touches #5662

**Comment by pieh** [comment]
> Same here, maybe gatsby should define a folder where such files should be cached. Not all downloaded files need to end up in public/static. Maybe .cache/files/[plugin-name]. This also touches #5662

We do have [createRemoteFileNode](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-filesystem#createremotefilenode) which handles downloads, caching but also creates file nodes which won't really work here - if this would be broken down so we could import just downloading and caching part this could be solution here

**Comment by axe312ger** [comment]
I created #5698 which should get https://deploy-preview-5659--using-contentful.netlify.com/image-api/ deployed successfully after it was merged :)

**Comment by pieh** [comment]
> So these are the Contentful image download duplicates:
> 
> * https://github.com/axe312ger/gatsby/blob/3b5d355bf739dc767505cd02f9280f4afa6c109a/packages/gatsby-source-contentful/src/extend-node-type.js#L288-L342
> * https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-sqip/src/extend-node-type.js#L198-L248
> * https://github.com/Khaledgarbaya/gatsby/blob/ba9cef15ee44be31c85c6f121fc2029bbe9b9c8f/packages/gatsby-remark-images-contentful/src/index.js#L45-L53

As Khaled pointed out in there - it doesn't actually download and save file, so we should skip it from discussion (my bad for bringing this up).

> I see two options:
> 
> 1. Make a generic function to download Contentful images and cache them outside of public/static
> 2. Make a generic function to download any file and cache them outside of public/static
>
> Both should return a Gatsby file object since in many cases these structure is needed. Others use cases can just extract the values they need.
> 
> What do you think?

As I mentioned above there is [createRemoteFileNode](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-filesystem#createremotefilenode) (used by some source plugins - like wordpress/drupal to download media library) - we could refactor it a bit so we could use just downloading and caching part of it. In situation like this we don't really want to create Gatsby File node because at this stage (doing work in resolver) schema is already created. It shouldn't break anything to just use it in its current form and extract path to downloaded file from it - it just would do some unnecessary work/overhead).

I can implement it first in `sqip` plugin and then you could "copy" this to this PR - would do you think about that?

**Comment by axe312ger** [comment]
@pieh sounds wonderful! We might consider just adding a new flag like `non-permanent` so it won't store it to the Gatsby db but does everything else. However, I am sure you will do it properly :)

File: packages/gatsby-source-contentful/src/extend-node-type.js [hunk 1]
```diff
@@ -265,6 +272,86 @@ exports.extendNodeType = ({ type }) => {
...
+    } = require(`gatsby-plugin-sharp`)
+
+    const { image, options } = args
+    const { id, file: { url, fileName } } = image
+    const { maxWidth, maxHeight, width, height, resizingBehavior, cropFocus, background } = options
+
+    // Downloading small version of the image with same aspect ratio
+    const assetWidth = maxWidth || width
+    const assetHeight = maxHeight || height
+    let aspectRatio = image.aspectRatio
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/extend-node-type.js]: I was testing this locally and `image.aspectRatio` was never defined for me - had to use this:
```
let aspectRatio = image.file.details.image.height / image.file.details.image.width
```
what am I missing here?

**Comment by pieh** [comment]
Just posting progress update - I started working on using `createRemoteFileNode` utility but hit road block: currently it doesn't handle query params, so will need to handle that first.

File: packages/gatsby-source-contentful/package.json [hunk 2]
```diff
@@ -33,5 +33,8 @@
     "build": "babel src --out-dir . --ignore __tests__",
     "prepublish": "cross-env NODE_ENV=production npm run build",
     "watch": "babel -w src --out-dir . --ignore __tests__"
+  },
+  "optionalDependencies": {
+    "gatsby-plugin-sharp": "^1.6.46"
```
Comment (pieh) [inline packages/gatsby-source-contentful/package.json]: If this is set as optional there should be some kind of check in code if it's available before using it. I think this should be just moved to `dependencies`

**Comment by pieh** [comment]
My current thinking is to merge this almost as-is so switching to using `createRemoteFileNode` don't block it. Only other changes I would like (not counting those in inline comments) is to not use `const { tmpdir } = require(`os`)` and download images to `.cache` rather than system tmp. Sounds good?

File: packages/gatsby-source-contentful/package.json (hunk 2 above)
Comment (axe312ger) [reply packages/gatsby-source-contentful/package.json]: Lets just move it, everybody working with images should have that one anyways.

I wonder if, at this point of time, there is even a scenario where you would not need `gatsby-plugin-sharp`. Move it & we are done here :)

**Comment by axe312ger** [comment]
Sounds good to me,

Switching to `.cache `makes sense, wanted to do the very same with the sqip images.

Why would you not use `const { tmpdir } = require(os)`?

File: packages/gatsby-transformer-sqip/src/extend-node-type.js [hunk 3]
```diff
@@ -42,7 +41,12 @@ module.exports = async args => {
   return {}
 }
 
-async function sqipSharp({ type, cache, getNodeAndSavePathDependency }) {
+async function sqipSharp({ type, cache, getNodeAndSavePathDependency, store }) {
+  const program = store.getState().program
+  const cacheDir = resolve(`${program.directory}/.cache/sqip/`)
```
Comment (axe312ger) [inline packages/gatsby-transformer-sqip/src/extend-node-type.js]: Hint: This changes the cacheDir of SQIP to be aligned with the cached Contentful assets. Both of these do not need to be publicly available.

**Comment by axe312ger** [comment]
@pieh @m-allanson @KyleAMathews This is ready for review.

IMHO the PR cleans up the code a lot and adds features while increasing readability :)

**Comment by pieh** [comment]
This looks good - I have just 2 semi related issues:
 - we need to split this into 2 PRs - move `using-contentful` changes to separate PR to avoid problems with preview and not yet published plugin (I can manage this one)
 - tracedSVG examples uses same images as default `sizes` example which mean images will be already loaded when user get to tracedSVG section and visitors won't be able to see tracedSVG placeholders - maybe we can merge them together? examples you added use `sizes` just with extra tracedSVG field

**Comment by axe312ger** [comment]
Done, I did split it up into two PRs. See #5859

**Comment by axe312ger** [comment]
I incremented the size by one pixel to ensure the user will see the traced SVG :) (In the other PR)

Anything else? 🤝

**Comment by axe312ger** [comment]
🎉

**Comment by nimaiwalsh** [comment]
Awesome work!

Sorry for being a n00b, but can someone please provide an example query to include an image sourced from Contentful using the traced image?

I have only used imageSharp with a local file like below in the past.
```   
imageSharp(id: { regex: "/bg.jpeg/" }) {
      sizes(maxWidth: 1240) {
        ...GatsbyImageSharpSizes
      }
  }
```
Many thanks

**Comment by nimaiwalsh** [comment]
@pieh  Awesome. Thanks.

**Pull request merged**